### PR TITLE
refactor: replace axios with native fetch across the monorepo

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
## Summary

Started as a full `axios` → native `fetch` migration across the monorepo. After review, the scope was narrowed: production HTTP call sites that needed explicit `undici` plumbing (a custom `dispatcher`, `EnvHttpProxyAgent`, tuned `Agent` pipelining) or otherwise carried a real risk of silent/implicit failures were reverted back to `axios`. `fetch` is kept only where it's a plain, dependency-free HTTP client with no special behavior to get right — i.e. in e2e test code.

- `@appium/support`'s `net.ts` (`uploadFile`/`downloadFile`), `@appium/base-driver`'s `WebDriverProxy`/`ProxyRequest` and `queryAppLink`, and `appium` core's GitHub-API version-lookup helper (`helpers/build.ts`) are back on `axios`, unchanged from `appium4`. `axios` remains a dependency of `@appium/support`, `@appium/base-driver`, and `appium`.
- e2e spec files that used `axios` purely as a generic HTTP test client (no special config) now use a small shared `fetch`-based helper (`httpGet`/`httpPost`/`httpDelete`/`httpRequest`) added to `@appium/driver-test-support`. These use Node's global `fetch` directly — no `undici` import anywhere in the repo.
- `net.ts` drops the `form-data` package: `axios`'s Node adapter builds the multipart request itself for any [spec-compliant `FormData`](https://github.com/axios/axios/blob/v1.x/lib/adapters/http.js), so `uploadFileToHttp`'s multipart path now builds the body with native `FormData` + `Blob` (`fs.openAsBlob()`) — both Node globals, no import required. Verified live against a local server that this produces byte-correct multipart output with a matching `Content-Length`.
  - Kept `mime-types` as a dependency to infer each part's `Content-Type` from its filename, since `form-data` did this internally and dropping it silently would send every part as `application/octet-stream`.
  - The non-multipart raw-upload path stays a plain `fs.createReadStream()`, not a `Blob`: `axios` force-overwrites a `Blob` body's `Content-Type` to `data.type || 'application/octet-stream'`, which would silently clobber a caller-supplied header. Verified live that a caller-supplied `Content-Type` now survives instead of being dropped.
- Fixed one latent bug uncovered along the way (kept even though the rest of `net.ts` reverted): `uploadFile()` set a raw-file-size `Content-Length` whenever `fileFieldName` was left unset, but the default is multipart — sending the wrong length for a differently-sized multipart body. Now only applies when the *effective* field name is falsy.

No public API changes. `WebDriverProxy`'s private `dispatcher`/`httpAgent`/`httpsAgent` fields and `ProxyRequest` were never part of the public surface.

### CODE REVIEW

**Comment by eglitise** [comment]
I think that since we're adding `undici` anyway, it might also be fine to use something like `ky` instead, which wraps `fetch` but adds convenience wrappers.
Edit: `ky` also has about 4x smaller install size than `undici`, so that's another reason to prefer it.

File: packages/driver-test-support/lib/helpers.ts [hunk 1]
```diff
@@ -5,6 +5,96 @@ import net from 'node:net';
...
+      if (!hasContentType) {
+        requestHeaders['content-type'] = 'application/x-www-form-urlencoded';
+      }
+    } else {
+      body = JSON.stringify(data);
+      if (!hasContentType) {
+        requestHeaders['content-type'] = 'application/json';
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/driver-test-support/lib/helpers.ts]: Fixed in 7074905bf — Buffer now hits the same branch as string and is sent as-is (with the same default content-type).

File: packages/support/lib/net.ts [hunk 2]
```diff
@@ -92,13 +81,15 @@ export async function uploadFile(
         `Only http/https protocols are supported.`,
     );
   }
-  if (!uploadOptions.fileFieldName) {
+  // Unset `fileFieldName` defaults to multipart; only an explicit falsy value means non-multipart,
+  // which is what this raw-file-size Content-Length applies to.
+  if ('fileFieldName' in uploadOptions && !uploadOptions.fileFieldName) {
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/support/lib/net.ts]: Fixed in 7074905bf — this now mirrors uploadFileToHttp()'s own `?? 'file'` default exactly, so an explicit `fileFieldName: undefined` no longer trips the raw-size Content-Length path.

File: packages/support/lib/net.ts [hunk 3]
```diff
@@ -121,32 +112,32 @@ export async function downloadFile(
...
+    const response = await fetch(remoteUrl, {
+      headers: toStringHeaders(requestHeaders),
+      signal: AbortSignal.timeout(timeout),
+    });
+    if (!response.ok) {
+      throw new Error(`Request failed with status code ${response.status}`);
+    }
+    responseLength = parseInt(response.headers.get('content-length') ?? '0', 10);
+    const responseStream = Readable.fromWeb(response.body as ReadableStream);
+    responseStream.pipe(writer);
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/support/lib/net.ts]: Fixed in 7074905bf — same fix as queryAppLink() (also fixed for null response.body in an earlier commit).

File: packages/base-driver/lib/wd-proxy/proxy-request.ts [hunk 4]
```diff
@@ -36,18 +50,52 @@ export class ProxyRequest {
...
+      }
+      if (err instanceof Error && 'response' in err) {
+        throw err;
       }
-      throw err;
+      // `fetch` wraps network errors (e.g. ECONNREFUSED) as a generic 'fetch failed' TypeError
+      // with the real reason in `.cause`; surface that reason in the message too.
+      const cause = (err as Error & {cause?: unknown})?.cause;
+      const reason = cause instanceof Error ? `: ${cause.message}` : '';
+      throw new Error(`${(err as Error).message}${reason}`, {cause: err});
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/base-driver/lib/wd-proxy/proxy-request.ts]: Fixed in 7074905bf — falls back to String(err) when the caught value isn't an Error instance.

File: packages/storage-plugin/test/e2e/storage.e2e.spec.ts [hunk 5]
```diff
@@ -30,6 +29,19 @@ const WDIO_OPTS: WebdriverIOConfig = {
...
+}
+
+async function postJson(url: string, data?: unknown): Promise<any> {
+  const response = await fetch(url, {
+    method: 'POST',
+    headers: typeof data === 'undefined' ? undefined : {'content-type': 'application/json'},
+    body: typeof data === 'undefined' ? undefined : JSON.stringify(data),
+  });
+  return response.json();
+}
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/storage-plugin/test/e2e/storage.e2e.spec.ts:50]: Fixed in 7074905bf — both helpers now throw on a non-2xx response.

**Comment by mykola-mokhnach** [comment]
> I think that since we're adding `undici` anyway, it might also be fine to use something like `ky` instead, which wraps `fetch` but adds convenience wrappers. Edit: `ky` also has about 4x smaller install size than `undici`, so that's another reason to prefer it.


### The core problem with the suggestion
`ky` wraps `fetch`; it doesn't implement its own connection layer. It can accept a `dispatcher` option and pass it through to the underlying call, but **you still need `undici`'s own `Agent`/`Dispatcher` classes to construct that value** — `ky` doesn't replace that need, it just sits on top of it. So swapping to `ky` at our 3 undici call sites would make `undici` an *additional* dependency alongside `ky`, not a replacement — the opposite of what "replacing undici imports with ky" implies.

### Pros
- Would centralize the "throw on non-2xx by default, `throwHttpErrors: false` to opt out" pattern we've hand-rolled in ~5 places (`net.ts`, `helpers.ts`, `driver-test-support`'s `httpRequest`, `storage-plugin`'s local helpers) into one tested library behavior — this is the actual duplication the earlier code review flagged.
- Built-in `timeout` option could replace our manual `AbortSignal.timeout()` wiring in a few spots.
- Zero deps, ESM-native, and its Node support floor is **Node 22+**, which happens to match this repo's own engine floor exactly — no compat-shim concerns.

### Cons
- **Doesn't remove undici** from the 3 places we actually need it (see above) — it'd be additive there, working against this PR's whole point (shedding a dependency after axios).
- **Wider dependency footprint**: to get the ergonomic benefit anywhere, `ky` would need adding to `@appium/support`, `appium` core, and `@appium/driver-test-support` at minimum — more packages than currently touch `undici` (just `@appium/base-driver` + one `fake-driver` test file).
- **Default retries are a real behavior-change risk**: `ky` retries GET/PUT/HEAD/DELETE/OPTIONS/TRACE/QUERY up to twice on network errors and 408/413/429/500/502/503/504. Silently retrying a proxied WebDriver GET command or an app download changes timing/side-effect assumptions unless `retry: {limit: 0}` is set at every call site — reintroducing exactly the kind of implicit magic this migration was trying to strip out.
- **`.json()` throws on a non-JSON body** rather than falling back to raw text. We specifically need the axios-style "try JSON.parse, fall back to string" behavior (several e2e tests rely on it, e.g. non-JSON error bodies), so we'd still hand-write that logic — the convenience gain is smaller than it looks.
- **Scope/timing**: this PR is already through a full review pass (CI green, Copilot comments addressed). Pulling in a new dependency now reopens review surface for a refactor that's orthogonal to "remove axios."

### Recommendation
Don't adopt `ky` in this PR. It can't replace `undici` where we actually need it, and where we don't need it (plain `fetch` call sites), the main win — deduping the throw/parse boilerplate — is small enough to do in-house as a ~15-line internal helper without a new dependency or the retry-semantics risk. Worth flagging as a possible separate, optional follow-up if the team wants it, not a blocker here.

File: packages/base-driver/lib/wd-proxy/proxy-request.ts [hunk 6]
```diff
@@ -36,18 +50,53 @@ export class ProxyRequest {
...
+  private async _makeRequest(signal: AbortSignal): Promise<ProxiedResponse> {
+    const {url, method, headers, data, timeout, dispatcher} = this._requestConfig;
     try {
-      return await axios({
-        ...this._requestConfig,
-        signal,
+      const timeoutSignal = timeout ? AbortSignal.timeout(timeout) : null;
+      const stringHeaders = headers
+        ? Object.fromEntries(Object.entries(headers).map(([key, value]) => [key, String(value)]))
+        : undefined;
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/base-driver/lib/wd-proxy/proxy-request.ts]: Fixed in 0a3c55d2c — nullish header values are now omitted and arrays are joined with ', ' like axios did.

File: packages/support/lib/net.ts [hunk 7]
```diff
@@ -176,7 +168,11 @@ export async function downloadFile(
   }
 }
 
-function toAxiosAuth(auth: AuthLike | undefined): AxiosBasicCredentials | null {
+function toStringHeaders(headers: Record<string, unknown>): Record<string, string> {
+  return Object.fromEntries(Object.entries(headers).map(([key, value]) => [key, String(value)]));
+}
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/support/lib/net.ts]: Fixed in 0a3c55d2c — toStringHeaders() now omits nullish values and joins arrays.

File: packages/support/lib/net.ts [hunk 8]
```diff
@@ -226,36 +217,39 @@ async function uploadFileToHttp(
       }
       for (const [key, value] of pairs) {
         if (key.toLowerCase() !== fileFieldName?.toLowerCase()) {
-          form.append(key, value as string | Buffer);
+          form.append(key, String(value));
         }
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/support/lib/net.ts:255]: Fixed in 0a3c55d2c — formFields now preserve string/Blob as-is and wrap a Buffer in a Blob, instead of coercing everything through String().

File: packages/base-driver/lib/wd-proxy/proxy.ts [hunk 9]
```diff
@@ -81,13 +78,12 @@ export class WebDriverProxy {
...
-    const agentOpts = {
-      keepAlive: opts.keepAlive ?? true,
-      maxSockets: 10,
-      maxFreeSockets: 5,
-    };
-    this.httpAgent = new http.Agent(agentOpts);
-    this.httpsAgent = new https.Agent(agentOpts);
+    this.dispatcher = new UndiciAgent({
+      connections: 10,
+      pipelining: 0,
```
Comment (KazuCocoa) [inline packages/base-driver/lib/wd-proxy/proxy.ts]: [P2] Keep keep-alive enabled by default

Undici defines `pipelining: 0` as disabling keep-alive entirely, so the default `keepAlive: true` option and both timeout settings below currently have no effect. Every proxied WebDriver command opens a fresh connection. With Undici 8.10.1 I confirmed four sequential requests used four TCP connections here, while `pipelining: 1` reused one. Please leave `pipelining` at 1 when keep-alive is enabled and use 0 only when the caller explicitly disables it.

File: packages/base-driver/lib/basedriver/helpers.ts [hunk 10]
```diff
@@ -426,27 +425,42 @@ function toCacheKey(app: string): string {
...
   try {
-    const {data: stream, headers, status} = await axios(requestOpts);
-    return {stream, headers, status};
+    // Omit null/undefined values (axios treats them as "unset"); join arrays like axios does.
+    const stringHeaders = Object.fromEntries(
+      Object.entries(headers)
+        .filter(([, value]) => value !== null && value !== undefined)
+        .map(([key, value]) => [key, Array.isArray(value) ? value.join(', ') : String(value)]),
+    );
+    const response = await fetch(requestUrl, {
```
Comment (KazuCocoa) [inline packages/base-driver/lib/basedriver/helpers.ts]: [P1] Preserve environment-proxy support

This replaces an Axios request that automatically honored `HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY`. Native `fetch` only does so when Node is started with `NODE_USE_ENV_PROXY=1`/`--use-env-proxy`, so existing installations behind corporate proxies can no longer download remote apps. The same regression affects the `fetch` calls in `support/lib/net.ts` and `appium/lib/helpers/build.ts`. I reproduced this with Axios 1.19.0 and the supported Node runtime: Axios reached a loopback proxy while `fetch` bypassed it and failed DNS resolution. Please use a proxy-aware dispatcher or shared wrapper that preserves the previous environment-variable behavior.

File: packages/support/lib/net.ts [hunk 11]
```diff
@@ -225,37 +224,47 @@ async function uploadFileToHttp(
...
-    form.append(fileFieldName, localFileStream);
-    requestOpts.headers = {
-      ...(isPlainObject(headers) ? headers : {}),
-      ...form.getHeaders(),
-    };
-    requestOpts.data = form;
-  } else {
-    if (isPlainObject(headers)) {
-      requestOpts.headers = headers;
+    form.append(fileFieldName, await openAsBlob(localPath), path.basename(localPath));
```
Comment (KazuCocoa) [inline packages/support/lib/net.ts]: [P2] Preserve the uploaded file MIME type

The previous `form-data` implementation inferred each file part MIME type from its filename. `openAsBlob()` is called without a `type`, so native `FormData` now emits `application/octet-stream` for every file. I confirmed the same PNG produces `image/png` before this change and `application/octet-stream` here. Servers or signed-upload policies that validate the part MIME type can reject previously valid uploads, and the stripped top-level `Content-Type` header cannot override a part. Please infer the MIME type and pass it through `openAsBlob(localPath, {type})`.

File: packages/base-driver/lib/wd-proxy/proxy.ts [hunk 12]
```diff
@@ -41,9 +39,8 @@ export class WebDriverProxy {
   sessionId: string | null;
   timeout: number;
   readonly headers: HTTPHeaders | undefined;
-  readonly httpAgent: http.Agent;
-  readonly httpsAgent: https.Agent;
 
+  private readonly dispatcher: UndiciAgent;
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/base-driver/lib/wd-proxy/proxy.ts]: Fixed the description in 328913283 — you're right, dispatcher is now private (not a public replacement for httpAgent/httpsAgent), so this is actually a smaller breaking-surface change than the description said. Updated the PR description to match.

File: packages/base-driver/lib/wd-proxy/proxy.ts (hunk 9 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/wd-proxy/proxy.ts]: Confirmed and fixed in 328913283 — verified live (4 sequential requests opened 4 TCP connections with pipelining: 0 vs. 2 with pipelining: 1). Now uses pipelining: 1 when keepAlive is enabled. Thanks for catching this.

File: packages/base-driver/lib/basedriver/helpers.ts (hunk 10 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.ts]: Confirmed and fixed in 328913283 — verified live that plain fetch bypasses HTTP_PROXY/HTTPS_PROXY entirely without --use-env-proxy. Added a fetchWithEnvProxy helper in @appium/support built on undici's EnvHttpProxyAgent and switched all three affected call sites to it. This also surfaced a second, more severe bug: uploadFileToHttp's multipart body was being built with the ambient global FormData, which undici's own fetch doesn't recognize as a FormData instance (same cross-undici-instance issue as the dispatcher/fetch mismatch elsewhere in this PR) — it was silently sending the literal string "[object FormData]" instead of real multipart bytes. Fixed by constructing the FormData via undici's own export.

File: packages/support/lib/net.ts (hunk 11 above)
Comment (mykola-mokhnach) [reply packages/support/lib/net.ts]: Fixed in 328913283 — file parts now get their MIME type inferred from the filename via mime-types and passed to openAsBlob().

File: packages/support/lib/net.ts [hunk 13]
```diff
@@ -176,7 +168,30 @@ export async function downloadFile(
...
+// #region Private helpers
+
+/**
+ * `fetch()` via undici's own `EnvHttpProxyAgent`, so `HTTP_PROXY`/`HTTPS_PROXY`/`NO_PROXY` are
+ * honored the way axios honored them by default (Node's global `fetch` ignores them unless
+ * started with `--use-env-proxy`).
+ */
+async function fetchWithEnvProxy(url: string, init: RequestInit = {}): Promise<Response> {
+  return await undiciFetch(url, {...init, dispatcher: new EnvHttpProxyAgent()});
+}
```
[the remark this answers is not in the document]
Comment (eglitise) [reply packages/support/lib/net.ts]: Note, there's 3 more suppressed Copilot comments under 'Review details' in the parent comment.

**Comment by eglitise** [comment]
Is it not safe to fully replace axios for `appium`, `fake-driver`, `support` and 2 script files, or do all of them still make use of the dispatcher functionality?

**Comment by mykola-mokhnach** [comment]
> Is it not safe to fully replace axios for `appium`, `fake-driver`, `support` and 2 script files, or do all of them still make use of the dispatcher functionality?

they all need the dispatcher. I've explicitly asked claude to only keep these that don't need undici imports

**Comment by eglitise** [comment]
...even the Crowdin sync and GitHub contribution stats scripts?

**Comment by mykola-mokhnach** [comment]
> ...even the Crowdin sync and GitHub contribution stats scripts?

only if we assume they don't need to respect local proxies

**Comment by eglitise** [comment]
They're not intended for end users and only run in CI, so yes, I would assume that, but fair enough

**Comment by mykola-mokhnach** [comment]
> They're not intended for end users and only run in CI, so yes, I would assume that, but fair enough

would you still prefer to migrate these scripts to use fetch?

**Comment by eglitise** [comment]
No, I guess it's fine

File: packages/driver-test-support/lib/helpers.ts [hunk 14]
```diff
@@ -5,6 +5,96 @@ import net from 'node:net';
...
+
+/**
+ * A minimal `fetch`-based HTTP test client with axios-like ergonomics:
+ * a `{status, data, headers}` result (JSON auto-parsed), and throw-by-default on non-2xx.
+ *
+ * Like axios, a string/Buffer `data` is sent as-is (no auto content-type); a plain
+ * object/array `data` is JSON-stringified with an `application/json` content-type added
+ * (unless already set).
+ */
+export async function httpRequest<T = any>(url: string, options: HttpRequestOptions = {}): Promise<HttpResult<T>> {
```
Comment (eglitise) [inline packages/driver-test-support/lib/helpers.ts]: There's many callers in the e2e test files importing `httpRequest`, but shouldn't they all use `httpGet`/`httpPost`/`httpDelete` instead? That way this method also won't need to be exported

Comment (mykola-mokhnach) [reply packages/driver-test-support/lib/helpers.ts]: Good call, done in 64e87486c - all callers now use `httpGet`/`httpPost`/`httpDelete`, and `httpRequest` is no longer exported (moved below them as a private implementation detail).

File: packages/support/lib/net.ts [hunk 15]
```diff
@@ -225,23 +232,33 @@ async function uploadFileToHttp(
...
       for (const [key, value] of pairs) {
-        if (key.toLowerCase() !== fileFieldName?.toLowerCase()) {
-          form.append(key, value as string | Buffer);
+        if (key.toLowerCase() === fileFieldName?.toLowerCase()) {
+          continue;
+        }
+        if (typeof value === 'string' || value instanceof Blob) {
+          form.append(key, value);
+        } else if (Buffer.isBuffer(value)) {
+          form.append(key, new Blob([value]));
```
Comment (KazuCocoa) [inline packages/support/lib/net.ts]: [P2] Preserve Buffer-valued form fields as fields

The previous `form-data` path serialized a Buffer-valued `formFields` entry with only `name="..."`; wrapping it in a Blob makes native FormData add `filename="blob"`. Multipart parsers therefore classify this value as a file instead of a normal field, so receivers reading it from their fields/body collection will stop seeing it. I verified the wire-format difference against `form-data@4.0.6`. Please preserve the non-file field semantics for Buffer values, or narrow and document the accepted `formFields` values as a breaking change.

Comment (mykola-mokhnach) [reply packages/support/lib/net.ts]: Confirmed live: native FormData always promotes a bare Blob to a File with `filename="blob"` on `.append()` - there's no spec-compliant way to send a Blob/Buffer as a non-file field. Since nothing in this repo actually passes a Buffer to `formFields`, went with your second option in c3bfdeadb: Buffer values now throw a clear TypeError instead of silently mis-serializing as a file attachment, and the JSDoc documents it.

**Review by KazuCocoa — APPROVED** [review]
lgtm
